File: kotlin/java-basic-example-vs-kotlin/src/main/java/org/example/ex06/Exam06.java
package org.example.ex06;

import lombok.Getter;

/**
 * @author : Rene
 * @since : 2023/10/22
 */
public class Exam06 {
	public Exam06(){
		var dog = new Dog("퍼피");
		dog.eat();
		dog.bark();
	}
	public static void main(String[] args) {
		new Exam06();
	}
}

@Getter
abstract class Animal implements Bark{
	private String name;
	public Animal(String name){
		this.name =name;
	}

	public void eat(){
		System.out.println("name = " + name);
	}
}


class Dog extends Animal{

	public Dog(){
		super("");
	}

	public Dog(String name){
		super(name);
	}

	@Override
	public void bark() {
		System.out.println(this.getName() + "bark");
	}
}

interface Bark{
	void bark();
}

File: kotlin/kotlin-basic-example/src/main/kotlin/ex06/Exam06.kt
package ex06

/**
 * @author        : Rene
 * @since         : 2023/10/22
 */
fun main() {
    // new 키워드 없이도 생성
    val dog = Dog("해피")

    dog.eat()
    dog.bark()
}


interface Bark {
    fun bark()
}

abstract class Animal (
    // 코틀린에서는 생성자 부분이 바로 선언이 됨
    private val name: String? = "") : Bark{

    // body - >

    fun eat() {
        println("$name 식사 !")
    }
}

// 이렇게만 선언해도 됨
//class Dog()


// Animal을 상속하는데 null을 받아도 되기 때문에 이렇게 호출이 가능함
//class Dog(): Animal()


// 생성자를 받아서 슈퍼라는 키워드로 넘기는 것  표현
class Dog(
    private val name:String?=null
) : Animal(name),
    // Temp 인터페이스 다중 상속 -> 파라미터 없으므로 그냥 상속
    Temp{
    override fun bark() {
        println("$name : 멍멍")
    }

    override fun hi() {
        println("hi")
    }
}

interface Temp{
    fun hi()
}

